sbncode/PID/Dazzle_module.cc

```diff
@@ -0,0 +1,810 @@
+{
+  art::ServiceHandle<cheat::BackTrackerService> bt_serv;
+  const int bestMatch(TruthMatchUtils::TrueParticleIDFromTotalTrueEnergy(clockData, hits, true));
+
+  if (!TruthMatchUtils::Valid(bestMatch))
+    return;
+
+  float totalHitEnergy(0.f), totalTrueHitEnergy(0.f);
+  for (auto const& hit : hits) {
+    const std::vector<sim::TrackIDE> trackIDEs(bt_serv->HitToTrackIDEs(clockData, hit));
+    totalHitEnergy = std::accumulate(trackIDEs.cbegin(), trackIDEs.cend(), totalHitEnergy,
+        [](float sum, auto const& ide) { return sum + ide.energy; });
+    totalTrueHitEnergy = std::accumulate(trackIDEs.cbegin(), trackIDEs.cend(), totalTrueHitEnergy,
+        [bestMatch](float sum, auto const& ide) { return (std::abs(ide.trackID) == bestMatch) ? sum + ide.energy : sum; });
+  }
+
+  const std::vector<const sim::IDE*> trackIDEs(bt_serv->TrackIdToSimIDEs_Ps(bestMatch));
+  float totalTrueEnergy(std::accumulate(trackIDEs.cbegin(), trackIDEs.cend(), 0.f,
+      [](float sum, auto const& ide) { return sum + ide->energy; }));
+
+  energyComp = totalTrueHitEnergy / totalTrueEnergy;
+  energyPurity = totalTrueHitEnergy / totalHitEnergy;
+
+  const simb::MCParticle* const trueParticle(particleInventory->TrackIdToParticle_P(bestMatch));
+
+  if (!trueParticle)
+    return;
+
+  truePdg = trueParticle->PdgCode();
+  trueType = this->PdgString(truePdg);
+  trueEndProcess = trueParticle->EndProcess();
+
+  trueStopping = (int)(trueEndProcess == "CoupledTransportation" || trueEndProcess == "FastScintillation" || trueEndProcess == "Decay" || trueEndProcess == "muMinusCaptureAtRest");
+
+  trueP = trueParticle->P();
+  trueEndP = trueParticle->P(trueParticle->NumberTrajectoryPoints() - 2);
+
+  const TVector3 trackStart(track.Start<TVector3>());
+  const TVector3 trackEnd(track.End<TVector3>());
+
+  const TVector3 trueStart(trueParticle->Position().Vect());
+  const TVector3 trueEnd(trueParticle->EndPosition().Vect());
+
+  trueStartX = trueStart.X();
+  trueStartY = trueStart.Y();
+  trueStartZ = trueStart.Z();
+
+  trueEndX = trueEnd.X();
+  trueEndY = trueEnd.Y();
+  trueEndZ = trueEnd.Z();
+
+  truePx = trueParticle->Px() / trueP;
+  truePy = trueParticle->Py() / trueP;
+  truePz = trueParticle->Pz() / trueP;
+
+  trueThetaXZ = std::atan2(truePx, truePz);
+  trueThetaYZ = std::atan2(truePy, truePz);
+
+  startDist = (trueStart - trackStart).Mag();
+  endDist = (trueEnd - trackEnd).Mag();
+}
+
+void Dazzle::FillTrackMetrics(const recob::Track& track)
+{
+  recoLen = track.Length();
+
+  const TVector3 start(track.Start<TVector3>());
+  const TVector3 end(track.End<TVector3>());
+  recoContained = this->InFV(start) && this->InFV(end);
+
+  if (!fMakeTree)
+    return;
+
+  startX = start.X();
+  startY = start.Y();
+  startZ = start.Z();
+
+  endX = end.X();
+  endY = end.Y();
+  endZ = end.Z();
+}
+
+bool Dazzle::InFV(const TVector3& pos) const
+{
+  return (std::abs(pos.X()) < 195 && std::abs(pos.Y()) < 195 && pos.Z() > 5 && pos.Z() < 495);
```

Comment:
This is just for setting a flag in the TTree right? I definitely think we shouldn't be skipping any reco steps based on FV considerations. Adjusting the FV is a classic analysis-time game.

Reply:
Yeah, this is so that you can select contained events for your training sample. The classification works on all tracks/showers so that the analyser can decide whether a track/shower is contained at a later point.

Reply:
This is now all configured via fcl

Reply:
I also suggest to go for "distance from the edges" in the configuration. That will also make it ICARUS-friendly.
GitHub is not showing me how you fixed this, so I can only ramble.

Reply:
I have pushed another commit that removes the symmetry assumptions that break for ICARUS, you can see the new implementation here: https://github.com/SBNSoftware/sbncode/blob/0620e981831a20a95506eda22d48becfa3f8491b/sbncode/PID/Razzle_module.cc#L516

Reply:
In the longer term, SBND has a very nice wrapper that provides a layer of abstraction that would allow me to continue to be ignorant towards the intricacies of the ICARUS geometry: https://github.com/SBNSoftware/sbndcode/blob/develop/sbndcode/Geometry/GeometryWrappers/TPCGeoAlg.h

I think this would need some modifications to be generalised to the multi-cryostat case before being ported to sbncode but I think it (or something similar ICARUS may already have?) is the best solution to avoid these issues (and general code duplication) in the long term. This is probably a more general discussion beyond the scope of this PR though.